File: 2022/03/03.py
import os


def get_priority(item):
    if item.lower() == item:
        return ord(item) - 96
    else:
        return ord(item) - 38


total = 0
with open(os.path.join(os.path.dirname(__file__), "03-input.txt")) as f:
    rucksacks = f.read().split("\n")
for rucksack in rucksacks:
    first_compartment = rucksack[0 : len(rucksack) // 2]
    second_compartment = rucksack[len(rucksack) // 2 :]
    for item in first_compartment:
        if item in second_compartment:
            total += get_priority(item)
            break

print("Part 1: " + str(total))

total = 0
counter = 0
while counter < len(rucksacks) - 1:
    for item in rucksacks[counter]:
        if item in rucksacks[counter + 1] and item in rucksacks[counter + 2]:
            total += get_priority(item)
            counter += 3
            break

print("Part 2: " + str(total))

File: 2022/03/03.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <array>
#include <set>
#include <algorithm>
#include <cctype>

int getPriority(char c) {
    if(std::islower(c)) {
        return c - 96;
    } else {
        return c - 38;
    }
}

int main(){
    std::string line;
    std::string rucksack;
    std::array<std::string, 300> rucksacks;
    std::array<char, 300> items;
    std::ifstream myfile("03-input.txt");
    if (!myfile) {
        std::cerr << "Unable to open file";
        exit(1);
    }
    int i = 0;
    while (getline(myfile, line)) {
        rucksack = line;
        rucksacks[i] = rucksack;
        i++;
    }
    myfile.close();
    i = 0;
    for (auto& r : rucksacks) {
        std::size_t mid = r.size() / 2;
        std::string first_half = r.substr(0, mid);
        std::string second_half = r.substr(mid);
        for (auto& c : first_half) {
            if(std::size_t found = second_half.find(c) != std::string::npos) {
                items[i] = c;
                i++;
                break;
            }
        }
    }
    int total = 0;
    for (auto& item : items) {
        total += getPriority(item);
    }
    std::cout << "Part 1: " << total << std::endl;

    i = 0;
    int badges = 0;
    while (i < 300) {
        for (auto& c : rucksacks[i]) {
            if (std::size_t found = rucksacks[i+1].find(c) != std::string::npos) {
                if (std::size_t found = rucksacks[i+2].find(c) != std::string::npos) {
                    badges += getPriority(c);
                    i += 3;
                    break;
                }
            }
        }

    }
       
    std::cout << "Part 2: " << badges << std::endl;
    return 0;
}